File: omi.py
#!/usr/bin/env python3
"""
Omi - Version Control for Python
Optimized Micro Index - Git-like commands with SQLite storage
"""

import sys
import os
import re
import sqlite3
import subprocess
import hashlib
import getpass
from datetime import datetime
from pathlib import Path
from io import BytesIO

# Try to import urllib3 for internal HTTP
try:
    import urllib3
    URLLIB3_AVAILABLE = True
except ImportError:
    URLLIB3_AVAILABLE = False


class Settings:
    """Load and manage settings from settings.txt"""
    
    @staticmethod
    def load():
        """Load settings from settings.txt file"""
        settings = {}
        
        if not os.path.exists("settings.txt"):
            print("Error: settings.txt not found")
            sys.exit(1)
        
        with open("settings.txt", "r") as f:
            for line in f:
                match = re.match(r"^([^=]+)=(.*)$", line.strip())
                if match:
                    key, value = match.groups()
                    settings[key] = value
        
        # Set defaults
        settings["API_ENABLED"] = settings.get("API_ENABLED", "1")
        settings["API_RATE_LIMIT"] = settings.get("API_RATE_LIMIT", "60")
        settings["API_RATE_LIMIT_WINDOW"] = settings.get("API_RATE_LIMIT_WINDOW", "60")
        
        return settings


class OmiRepository:
    """Manage Omi repository operations"""
    
    def __init__(self, settings):
        self.settings = settings
        self.db_name = self._read_dotomi() or "repo.omi"
    
    def _read_dotomi(self):
        """Read database name from .omi file"""
        if os.path.exists(".omi"):
            with open(".omi", "r") as f:
                match = re.search(r'OMI_DB="([^"]+)"', f.read())
                if match:
                    return match.group(1)
        return None
    
    def _write_dotomi(self, db_name):
        """Write database name to .omi file"""
        with open(".omi", "w") as f:
            f.write(f'OMI_DB="{db_name}"\n')
    
    def init(self, db_name="repo.omi"):
        """Initialize a new repository"""
        print("Initializing omi repository...")
        self.db_name = db_name
        self._write_dotomi(db_name)
        
        # Create database and tables
        conn = sqlite3.connect(db_name)
        cursor = conn.cursor()
        
        cursor.execute("""
            CREATE TABLE IF NOT EXISTS blobs (
                hash TEXT PRIMARY KEY,
                data BLOB,
                size INTEGER
            )
        """)
        
        cursor.execute("""
            CREATE TABLE IF NOT EXISTS files (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                filename TEXT,
                hash TEXT,
                datetime TEXT,
                commit_id INTEGER,
                FOREIGN KEY(commit_id) REFERENCES commits(id)
            )
        """)
        
        cursor.execute("""
            CREATE TABLE IF NOT EXISTS commits (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                message TEXT,
                datetime TEXT,
                user TEXT
            )
        """)
        
        cursor.execute("""
            CREATE TABLE IF NOT EXISTS staging (
                filename TEXT PRIMARY KEY,
                hash TEXT,
                datetime TEXT
            )
        """)
        
        cursor.execute("CREATE INDEX IF NOT EXISTS idx_files_hash ON files(hash)")
        cursor.execute("CREATE INDEX IF NOT EXISTS idx_files_commit ON files(commit_id)")
        cursor.execute("CREATE INDEX IF NOT EXISTS idx_blobs_hash ON blobs(hash)")
        
        conn.commit()
        conn.close()
        
        print(f"Repository initialized: {db_name}")
    
    def clone(self, url):
        """Clone a repository from local or remote source"""
        print(f"Cloning from {url}...")
        
        if os.path.isfile(url):
            # Local clone
            import shutil
            shutil.copy(url, "repo.omi")
            self._write_dotomi("repo.omi")
            self.db_name = "repo.omi"
            print("Cloned to repo.omi")
        else:
            # Remote clone
            repo_name = os.path.basename(url) or "repo.omi"
            print(f"Downloading {repo_name} from {self.settings['REPOS']}...")
            
            cmd = [
                self.settings["CURL"],
                "-f", "-o", repo_name,
                f"{self.settings['REPOS']}/?download={repo_name}"
            ]
            
            if subprocess.run(cmd, capture_output=True).returncode == 0:
                self._write_dotomi(repo_name)
                self.db_name = repo_name
                print(f"Cloned to {repo_name}")
            else:
                print("Error: Failed to clone from remote")
                sys.exit(1)
    
    def add_files(self, pattern="--all"):
        """Stage files for commit"""
        print("Adding files to staging...")
        
        if pattern == "--all":
            # Add all files in current directory
            for filename in os.listdir("."):
                if (os.path.isfile(filename) and 
                    filename != self.db_name and 
                    filename != ".omi"):
                    self._add_one_file(filename)
        else:
            self._add_one_file(pattern)
    
    def _add_one_file(self, filename):
        """Add a single file to staging"""
        if not os.path.isfile(filename):
            print(f"Error: File not found: {filename}")
            return False
        
        # Calculate SHA256 hash
        sha256_hash = hashlib.sha256()
        with open(filename, "rb") as f:
            sha256_hash.update(f.read())
        hash_value = sha256_hash.hexdigest()
        
        # Get current datetime
        datetime_str = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
        
        # Add to staging
        conn = sqlite3.connect(self.db_name)
        cursor = conn.cursor()
        cursor.execute(
            "INSERT OR REPLACE INTO staging (filename, hash, datetime) VALUES (?, ?, ?)",
            (filename, hash_value, datetime_str)
        )
        conn.commit()
        conn.close()
        
        print(f"Staged: {filename} (hash: {hash_value})")
        return True
    
    def commit(self, message="No message"):
        """Create a commit from staged files"""
        print("Committing changes...")
        
        datetime_str = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
        user = os.getenv("USER", "unknown")
        
        conn = sqlite3.connect(self.db_name)
        cursor = conn.cursor()
        
        # Create commit record
        cursor.execute(
            "INSERT INTO commits (message, datetime, user) VALUES (?, ?, ?)",
            (message, datetime_str, user)
        )
        commit_id = cursor.lastrowid
        
        # Get staged files
        cursor.execute("SELECT filename, hash, datetime FROM staging")
        staged_files = cursor.fetchall()
        
        # Process each staged file
        for filename, hash_value, file_datetime in staged_files:
            self._commit_one_file(cursor, filename, hash_value, file_datetime, commit_id)
        
        # Clear staging
        cursor.execute("DELETE FROM staging")
        
        conn.commit()
        conn.close()
        
        print(f"Committed successfully (commit #{commit_id})")
    
    def _commit_one_file(self, cursor, filename, hash_value, file_datetime, commit_id):
        """Add a single file to a commit"""
        # Check if blob exists (deduplication)
        cursor.execute("SELECT COUNT(*) FROM blobs WHERE hash=?", (hash_value,))
        blob_count = cursor.fetchone()[0]
        
        if blob_count == 0:
            # Store new blob
            with open(filename, "rb") as f:
                data = f.read()
            
            cursor.execute(
                "INSERT INTO blobs (hash, data, size) VALUES (?, ?, ?)",
                (hash_value, data, len(data))
            )
            print(f"  Stored new blob: {hash_value}")
        else:
            print(f"  Blob already exists (deduplicated): {hash_value}")
        
        # Add file record
        cursor.execute(
            "INSERT INTO files (filename, hash, datetime, commit_id) VALUES (?, ?, ?, ?)",
            (filename, hash_value, file_datetime, commit_id)
        )
    
    def _http_post_multipart(self, url, fields):
        """Upload files using multipart form data (urllib3 if available, else curl)"""
        use_internal = self.settings.get("USE_INTERNAL_HTTP", "1") == "1"
        
        if use_internal and URLLIB3_AVAILABLE:
            try:
                http = urllib3.PoolManager()
                response = http.request(
                    'POST',
                    url,
                    fields=fields,
                    timeout=float(self.settings.get("HTTP_TIMEOUT", "30"))
                )
                return response.status, response.data
            except Exception as e:
                print(f"Warning: Internal HTTP failed, falling back to curl: {e}")
                return self._http_post_multipart_curl(url, fields)
        else:
            return self._http_post_multipart_curl(url, fields)
    
    def _http_post_multipart_curl(self, url, fields):
        """Upload files using curl executable"""
        cmd = [self.settings["CURL"], "-f", "-X", "POST"]
        
        for key, value in fields.items():
            if isinstance(value, tuple) and len(value) == 2:
                # File upload
                cmd.extend(["-F", f"{key}=@{value[1]}"])
            else:
                # Regular form field
                cmd.extend(["-F", f"{key}={value}"])
        
        cmd.append(url)
        
        result = subprocess.run(cmd, capture_output=True)
        return result.returncode, result.stdout
    
    def push(self):
        """Upload repository to remote server"""
        print(f"Pushing {self.db_name} to remote...")
        
        if not os.path.isfile(self.db_name):
            print(f"Error: Database file {self.db_name} not found")
            sys.exit(1)
        
        if self.settings["API_ENABLED"] == "0":
            print("Error: API is disabled")
            sys.exit(1)
        
        otp_code = ""
        if self._has_2fa_enabled():
            otp_code = getpass.getpass("Enter OTP code (6 digits): ")
        
        # Build form fields
        fields = {
            'username': self.settings['USERNAME'],
            'password': self.settings['PASSWORD'],
            'repo_name': self.db_name,
            'repo_file': ('file', self.db_name),
            'action': 'Upload'
        }
        
        if otp_code:
            fields['otp_code'] = otp_code
        
        # Use internal or external HTTP based on settings
        use_internal = self.settings.get("USE_INTERNAL_HTTP", "1") == "1"
        
        if use_internal and URLLIB3_AVAILABLE:
            try:
                status, response = self._http_post_multipart(
                    f"{self.settings['REPOS']}/",
                    fields
                )
                if status == 200:
                    print(f"Successfully pushed to {self.settings['REPOS']}")
                else:
                    print(f"Error: Server returned status {status}")
                    sys.exit(1)
            except Exception as e:
                print(f"Error: Failed to push: {e}")
                sys.exit(1)
        else:
            # Fall back to curl
            cmd = [
                self.settings["CURL"],
                "-f", "-X", "POST",
                "-F", f"username={self.settings['USERNAME']}",
                "-F", f"password={self.settings['PASSWORD']}",
                "-F", f"repo_name={self.db_name}",
                "-F", f"repo_file=@{self.db_name}",
                "-F", "action=Upload"
            ]
            
            if otp_code:
                cmd.extend(["-F", f"otp_code={otp_code}"])
            
            cmd.append(f"{self.settings['REPOS']}/")
            
            result = subprocess.run(cmd, capture_output=True, text=True)
            
            if result.returncode == 0:
                print(f"Successfully pushed to {self.settings['REPOS']}")
            else:
                print("Error: Failed to push to remote")
                if result.stderr:
                    print(result.stderr)
                sys.exit(1)
    
    def pull(self):
        """Download repository from remote server"""
        print(f"Pulling {self.db_name} from remote...")
        
        if self.settings["API_ENABLED"] == "0":
            print("Error: API is disabled")
            sys.exit(1)
        
        otp_code = ""
        if self._has_2fa_enabled():
            otp_code = getpass.getpass("Enter OTP code (6 digits): ")
        
        use_internal = self.settings.get("USE_INTERNAL_HTTP", "1") == "1"
        
        if use_internal and URLLIB3_AVAILABLE:
            try:
                http = urllib3.PoolManager()
                fields = {
                    'username': self.settings['USERNAME'],
                    'password': self.settings['PASSWORD'],
                    'repo_name': self.db_name,
                    'action': 'pull'
                }
                if otp_code:
                    fields['otp_code'] = otp_code
                
                response = http.request(
                    'POST',
                    f"{self.settings['REPOS']}/",
                    fields=fields,
                    timeout=float(self.settings.get("HTTP_TIMEOUT", "30"))
                )
                
                if response.status == 200:
                    with open(self.db_name, "wb") as f:
                        f.write(response.data)
                    print(f"Successfully pulled from {self.settings['REPOS']}")
                else:
                    print(f"Error: Server returned status {response.status}")
                    sys.exit(1)
            except Exception as e:
                print(f"Error: Failed to pull: {e}")
                sys.exit(1)
        else:
            # Fall back to curl
            cmd = [
                self.settings["CURL"],
                "-f", "-X", "POST",
                "-d", f"username={self.settings['USERNAME']}",
                "-d", f"password={self.settings['PASSWORD']}",
                "-d", f"repo_name={self.db_name}",
                "-d", "action=pull"
            ]
            
            if otp_code:
                cmd.extend(["-d", f"otp_code={otp_code}"])
            
            cmd.append(f"{self.settings['REPOS']}/")
            
            result = subprocess.run(cmd, capture_output=True)
            
            if result.returncode == 0:
                # Write downloaded repository
                with open(self.db_name, "wb") as f:
                    f.write(result.stdout)
                print(f"Successfully pulled from {self.settings['REPOS']}")
            else:
                print("Error: Failed to pull from remote")
                if result.stderr:
                    print(result.stderr.decode())
                sys.exit(1)
    
    def list_repos(self):
        """List available repositories on remote server"""
        print(f"=== Available Repositories on {self.settings['REPOS']} ===")
        
        cmd = [
            self.settings["CURL"],
            "-s",
            f"{self.settings['REPOS']}/?format=json"
        ]
        
        result = subprocess.run(cmd, capture_output=True, text=True)
        
        if result.returncode == 0:
            # Extract repo names from JSON
            for match in re.finditer(r'"name":"([^"]*)"', result.stdout):
                print(match.group(1))
        else:
            print("Error: Failed to retrieve repository list")
            sys.exit(1)
    
    def show_status(self):
        """Show repository status"""
        print("=== Staged Files ===")
        self._execute_query("SELECT filename, datetime FROM staging")
        print("")
        
        print("=== Recent Commits ===")
        self._execute_query("SELECT id, message, datetime FROM commits ORDER BY id DESC LIMIT 5")
        print("")
        
        print("=== Statistics ===")
        result = self._execute_query("SELECT COUNT(*) FROM blobs", fetch_one=True)
        print(f"Total blobs (deduplicated): {result[0] if result else 0}")
        
        result = self._execute_query("SELECT COUNT(*) FROM files", fetch_one=True)
        print(f"Total file versions: {result[0] if result else 0}")
    
    def log_commits(self, limit=10):
        """Show commit history"""
        print("=== Commit History ===")
        query = f"SELECT id, datetime, user, message FROM commits ORDER BY id DESC LIMIT {limit}"
        self._execute_query(query)
    
    def _execute_query(self, query, fetch_one=False):
        """Execute a query and print results"""
        conn = sqlite3.connect(self.db_name)
        cursor = conn.cursor()
        cursor.execute(query)
        
        if fetch_one:
            return cursor.fetchone()
        
        rows = cursor.fetchall()
        for row in rows:
            print("|".join(str(col) for col in row))
        
        conn.close()
    
    def _has_2fa_enabled(self):
        """Check if user has 2FA enabled"""
        if not os.path.isfile("phpusers.txt"):
            return False
        
        username = self.settings.get("USERNAME", "")
        with open("phpusers.txt", "r") as f:
            for line in f:
                parts = line.strip().split(":")
                if len(parts) >= 3 and parts[0] == username and parts[2]:
                    return True
        
        return False


def main():
    """Main entry point"""
    settings = Settings.load()
    repo = OmiRepository(settings)
    
    if len(sys.argv) < 2:
        print("Usage: omi <command> [args]")
        print("Commands: init, clone, add, commit, push, pull, list, log, status")
        sys.exit(1)
    
    cmd = sys.argv[1]
    args = sys.argv[2:]
    
    try:
        if cmd == "init":
            db_name = args[0] if args else "repo.omi"
            repo.init(db_name)
        elif cmd == "clone":
            if not args:
                print("Usage: omi clone <url>")
                sys.exit(1)
            repo.clone(args[0])
        elif cmd == "add":
            pattern = args[0] if args else "--all"
            repo.add_files(pattern)
        elif cmd == "commit":
            message = "No message"
            # Parse -m "message"
            for i, arg in enumerate(args):
                if arg == "-m" and i + 1 < len(args):
                    message = args[i + 1]
                    break
            repo.commit(message)
        elif cmd == "push":
            repo.push()
        elif cmd == "pull":
            repo.pull()
        elif cmd == "list":
            repo.list_repos()
        elif cmd == "log":
            limit = int(args[0]) if args else 10
            repo.log_commits(limit)
        elif cmd == "status":
            repo.show_status()
        else:
            print(f"Unknown command: {cmd}")
            print("Usage: omi <command> [args]")
            print("Commands: init, clone, add, commit, push, pull, list, log, status")
            sys.exit(1)
    except KeyboardInterrupt:
        print("\nInterrupted by user")
        sys.exit(1)
    except Exception as e:
        print(f"Error: {e}")
        sys.exit(1)


if __name__ == "__main__":
    main()

File: omi.cs
using System;
using System.IO;
using System.Data;
using System.Data.SQLite;
using System.Text;
using System.Security.Cryptography;
using System.Collections.Generic;
using System.Diagnostics;
using System.Net;
using System.Net.Http;
using System.Threading.Tasks;

class Settings
{
    public string ServerUrl { get; set; }
    public string ApiKey { get; set; }
    public string OTPSecret { get; set; }
    public string LocalRepoPath { get; set; }
    public string Username { get; set; }
    public string Password { get; set; }
    public string Repos { get; set; }
    public bool UseInternalHttp { get; set; } = true;
    public int HttpTimeout { get; set; } = 30;

    public void LoadFromFile(string path)
    {
        if (!File.Exists(path)) return;
        
        var lines = File.ReadAllLines(path);
        foreach (var line in lines)
        {
            var trimmedLine = line.Trim();
            if (trimmedLine.StartsWith("#")) continue;
            
            if (trimmedLine.StartsWith("SERVER_URL="))
                ServerUrl = trimmedLine.Substring("SERVER_URL=".Length);
            else if (trimmedLine.StartsWith("API_KEY="))
                ApiKey = trimmedLine.Substring("API_KEY=".Length);
            else if (trimmedLine.StartsWith("OTP_SECRET="))
                OTPSecret = trimmedLine.Substring("OTP_SECRET=".Length);
            else if (trimmedLine.StartsWith("LOCAL_REPO_PATH="))
                LocalRepoPath = trimmedLine.Substring("LOCAL_REPO_PATH=".Length);
            else if (trimmedLine.StartsWith("USERNAME="))
                Username = trimmedLine.Substring("USERNAME=".Length);
            else if (trimmedLine.StartsWith("PASSWORD="))
                Password = trimmedLine.Substring("PASSWORD=".Length);
            else if (trimmedLine.StartsWith("REPOS="))
                Repos = trimmedLine.Substring("REPOS=".Length);
            else if (trimmedLine.StartsWith("USE_INTERNAL_HTTP="))
                UseInternalHttp = trimmedLine.Substring("USE_INTERNAL_HTTP=".Length) == "1";
            else if (trimmedLine.StartsWith("HTTP_TIMEOUT="))
                HttpTimeout = int.Parse(trimmedLine.Substring("HTTP_TIMEOUT=".Length));
        }
    }
}

class OmiRepository
{
    private Settings settings;
    private string dbPath;

    public OmiRepository(Settings settings, string dbPath)
    {
        this.settings = settings;
        this.dbPath = dbPath;
    }

    public void Init()
    {
        if (File.Exists(dbPath))
        {
            Console.WriteLine("Repository already initialized");
            return;
        }

        using (var connection = new SQLiteConnection($"Data Source={dbPath}"))
        {
            connection.Open();
            using (var cmd = connection.CreateCommand())
            {
                cmd.CommandText = @"
                    CREATE TABLE blobs (sha256 TEXT PRIMARY KEY, data BLOB);
                    CREATE TABLE files (path TEXT PRIMARY KEY, sha256 TEXT, modified DATETIME);
                    CREATE TABLE commits (hash TEXT PRIMARY KEY, message TEXT, files TEXT, timestamp DATETIME);
                    CREATE TABLE staging (path TEXT PRIMARY KEY, sha256 TEXT);
                ";
                cmd.ExecuteNonQuery();
            }
        }

        Console.WriteLine("Repository initialized");
    }

    public void AddFile(string filePath)
    {
        if (!File.Exists(filePath))
        {
            Console.WriteLine($"File not found: {filePath}");
            return;
        }

        var fileData = File.ReadAllBytes(filePath);
        var sha256 = ComputeSHA256(fileData);

        using (var connection = new SQLiteConnection($"Data Source={dbPath}"))
        {
            connection.Open();

            // Add blob
            using (var cmd = connection.CreateCommand())
            {
                cmd.CommandText = "INSERT OR IGNORE INTO blobs (sha256, data) VALUES (@sha, @data)";
                cmd.Parameters.AddWithValue("@sha", sha256);
                cmd.Parameters.AddWithValue("@data", fileData);
                cmd.ExecuteNonQuery();
            }

            // Add to staging
            using (var cmd = connection.CreateCommand())
            {
                cmd.CommandText = "INSERT OR REPLACE INTO staging (path, sha256) VALUES (@path, @sha)";
                cmd.Parameters.AddWithValue("@path", filePath);
                cmd.Parameters.AddWithValue("@sha", sha256);
                cmd.ExecuteNonQuery();
            }
        }

        Console.WriteLine($"Added: {filePath}");
    }

    public void Commit(string message)
    {
        using (var connection = new SQLiteConnection($"Data Source={dbPath}"))
        {
            connection.Open();

            // Get staged files
            var stagedFiles = new Dictionary<string, string>();
            using (var cmd = connection.CreateCommand())
            {
                cmd.CommandText = "SELECT path, sha256 FROM staging";
                using (var reader = cmd.ExecuteReader())
                {
                    while (reader.Read())
                    {
                        stagedFiles[reader.GetString(0)] = reader.GetString(1);
                    }
                }
            }

            if (stagedFiles.Count == 0)
            {
                Console.WriteLine("Nothing to commit");
                return;
            }

            // Create commit
            var timestamp = DateTime.UtcNow;
            var commitHash = ComputeSHA256(Encoding.UTF8.GetBytes(
                message + timestamp.Ticks.ToString())).Substring(0, 8);

            var filesList = string.Join(",", stagedFiles.Keys);

            using (var cmd = connection.CreateCommand())
            {
                cmd.CommandText = "INSERT INTO commits (hash, message, files, timestamp) VALUES (@hash, @msg, @files, @ts)";
                cmd.Parameters.AddWithValue("@hash", commitHash);
                cmd.Parameters.AddWithValue("@msg", message);
                cmd.Parameters.AddWithValue("@files", filesList);
                cmd.Parameters.AddWithValue("@ts", timestamp);
                cmd.ExecuteNonQuery();
            }

            // Update files table
            foreach (var kvp in stagedFiles)
            {
                using (var cmd = connection.CreateCommand())
                {
                    cmd.CommandText = "INSERT OR REPLACE INTO files (path, sha256, modified) VALUES (@path, @sha, @ts)";
                    cmd.Parameters.AddWithValue("@path", kvp.Key);
                    cmd.Parameters.AddWithValue("@sha", kvp.Value);
                    cmd.Parameters.AddWithValue("@ts", timestamp);
                    cmd.ExecuteNonQuery();
                }
            }

            // Clear staging
            using (var cmd = connection.CreateCommand())
            {
                cmd.CommandText = "DELETE FROM staging";
                cmd.ExecuteNonQuery();
            }

            Console.WriteLine($"Committed: {commitHash}");
        }
    }

    public void Push(string otpCode)
    {
        if (string.IsNullOrEmpty(settings.Repos))
        {
            Console.WriteLine("Server URL not configured");
            return;
        }

        if (!ValidateOTP(otpCode))
        {
            Console.WriteLine("Invalid OTP code");
            return;
        }

        if (!File.Exists(dbPath))
        {
            Console.WriteLine($"Error: Database file {dbPath} not found");
            return;
        }

        if (settings.UseInternalHttp)
        {
            try
            {
                PushWithHttpClient(otpCode);
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Internal HTTP failed: {ex.Message}, falling back to curl");
                PushWithCurl(otpCode);
            }
        }
        else
        {
            PushWithCurl(otpCode);
        }
    }

    private void PushWithHttpClient(string otpCode)
    {
        using (var client = new HttpClient())
        {
            client.Timeout = TimeSpan.FromSeconds(settings.HttpTimeout);

            using (var form = new MultipartFormDataContent())
            {
                form.Add(new StringContent(settings.Username), "username");
                form.Add(new StringContent(settings.Password), "password");
                form.Add(new StringContent(dbPath), "repo_name");
                form.Add(new StringContent("Upload"), "action");
                
                if (!string.IsNullOrEmpty(otpCode))
                    form.Add(new StringContent(otpCode), "otp_code");

                using (var fs = File.OpenRead(dbPath))
                {
                    form.Add(new StreamContent(fs), "repo_file", Path.GetFileName(dbPath));
                    
                    var response = client.PostAsync(settings.Repos + "/", form).Result;
                    
                    if (response.IsSuccessStatusCode)
                    {
                        Console.WriteLine($"Successfully pushed to {settings.Repos}");
                    }
                    else
                    {
                        Console.WriteLine($"Error: Server returned {response.StatusCode}");
                    }
                }
            }
        }
    }

    private void PushWithCurl(string otpCode)
    {
        var cmd = new List<string>
        {
            "curl", "-f", "-X", "POST",
            "-F", $"username={settings.Username}",
            "-F", $"password={settings.Password}",
            "-F", $"repo_name={dbPath}",
            "-F", $"repo_file=@{dbPath}",
            "-F", "action=Upload"
        };

        if (!string.IsNullOrEmpty(otpCode))
            cmd.AddRange(new[] { "-F", $"otp_code={otpCode}" });

        cmd.Add(settings.Repos + "/");

        var process = new Process
        {
            StartInfo = new ProcessStartInfo
            {
                FileName = "curl",
                Arguments = string.Join(" ", cmd.Skip(1)),
                UseShellExecute = false,
                RedirectStandardOutput = true,
                RedirectStandardError = true
            }
        };

        process.Start();
        var error = process.StandardError.ReadToEnd();
        process.WaitForExit();

        if (process.ExitCode == 0)
        {
            Console.WriteLine($"Successfully pushed to {settings.Repos}");
        }
        else
        {
            Console.WriteLine("Error: Failed to push to remote");
            if (!string.IsNullOrEmpty(error))
                Console.WriteLine(error);
        }
    }

    public void Pull(string otpCode)
    {
        if (string.IsNullOrEmpty(settings.Repos))
        {
            Console.WriteLine("Server URL not configured");
            return;
        }

        if (!ValidateOTP(otpCode))
        {
            Console.WriteLine("Invalid OTP code");
            return;
        }

        if (settings.UseInternalHttp)
        {
            try
            {
                PullWithHttpClient(otpCode);
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Internal HTTP failed: {ex.Message}, falling back to curl");
                PullWithCurl(otpCode);
            }
        }
        else
        {
            PullWithCurl(otpCode);
        }
    }

    private void PullWithHttpClient(string otpCode)
    {
        using (var client = new HttpClient())
        {
            client.Timeout = TimeSpan.FromSeconds(settings.HttpTimeout);

            var form = new Dictionary<string, string>
            {
                { "username", settings.Username },
                { "password", settings.Password },
                { "repo_name", dbPath },
                { "action", "pull" }
            };

            if (!string.IsNullOrEmpty(otpCode))
                form["otp_code"] = otpCode;

            var content = new FormUrlEncodedContent(form);
            var response = client.PostAsync(settings.Repos + "/", content).Result;

            if (response.IsSuccessStatusCode)
            {
                var bytes = response.Content.ReadAsByteArrayAsync().Result;
                File.WriteAllBytes(dbPath, bytes);
                Console.WriteLine($"Successfully pulled from {settings.Repos}");
            }
            else
            {
                Console.WriteLine($"Error: Server returned {response.StatusCode}");
            }
        }
    }

    private void PullWithCurl(string otpCode)
    {
        var cmd = new List<string>
        {
            "curl", "-f", "-X", "POST",
            "-d", $"username={settings.Username}",
            "-d", $"password={settings.Password}",
            "-d", $"repo_name={dbPath}",
            "-d", "action=pull"
        };

        if (!string.IsNullOrEmpty(otpCode))
            cmd.AddRange(new[] { "-d", $"otp_code={otpCode}" });

        cmd.Add(settings.Repos + "/");

        var process = new Process
        {
            StartInfo = new ProcessStartInfo
            {
                FileName = "curl",
                Arguments = string.Join(" ", cmd.Skip(1)),
                UseShellExecute = false,
                RedirectStandardOutput = true,
                RedirectStandardError = true
            }
        };

        process.Start();
        var output = process.StandardOutput.BaseStream;
        var error = process.StandardError.ReadToEnd();
        
        if (process.WaitForExit(settings.HttpTimeout * 1000))
        {
            if (process.ExitCode == 0)
            {
                using (var fs = File.OpenWrite(dbPath))
                {
                    output.CopyTo(fs);
                }
                Console.WriteLine($"Successfully pulled from {settings.Repos}");
            }
            else
            {
                Console.WriteLine("Error: Failed to pull from remote");
                if (!string.IsNullOrEmpty(error))
                    Console.WriteLine(error);
            }
        }
    }

    public void ListRepositories()
    {
        Console.WriteLine("Repositories:");
        if (File.Exists(dbPath))
            Console.WriteLine($"  {Path.GetFileName(dbPath)}");
    }

    public void ShowStatus()
    {
        if (!File.Exists(dbPath))
        {
            Console.WriteLine("Repository not initialized");
            return;
        }

        using (var connection = new SQLiteConnection($"Data Source={dbPath}"))
        {
            connection.Open();

            Console.WriteLine("Staged files:");
            using (var cmd = connection.CreateCommand())
            {
                cmd.CommandText = "SELECT path FROM staging";
                using (var reader = cmd.ExecuteReader())
                {
                    while (reader.Read())
                        Console.WriteLine($"  {reader.GetString(0)}");
                }
            }
        }
    }

    public void ShowLog()
    {
        if (!File.Exists(dbPath))
        {
            Console.WriteLine("Repository not initialized");
            return;
        }

        using (var connection = new SQLiteConnection($"Data Source={dbPath}"))
        {
            connection.Open();

            using (var cmd = connection.CreateCommand())
            {
                cmd.CommandText = "SELECT hash, message, timestamp FROM commits ORDER BY timestamp DESC";
                using (var reader = cmd.ExecuteReader())
                {
                    while (reader.Read())
                    {
                        Console.WriteLine($"[{reader.GetString(0)}] {reader.GetString(1)} ({reader.GetDateTime(2)})");
                    }
                }
            }
        }
    }

    private string ComputeSHA256(byte[] data)
    {
        using (var sha256 = SHA256.Create())
        {
            var hash = sha256.ComputeHash(data);
            var sb = new StringBuilder();
            foreach (var b in hash)
                sb.Append(b.ToString("x2"));
            return sb.ToString();
        }
    }

    private bool ValidateOTP(string code)
    {
        if (string.IsNullOrEmpty(settings.OTPSecret))
            return true;

        if (string.IsNullOrEmpty(code))
            return false;

        long timestamp = DateTimeOffset.UtcNow.ToUnixTimeSeconds() / 30;
        var expectedCode = GenerateTOTP(settings.OTPSecret, timestamp);

        return code == expectedCode;
    }

    private string GenerateTOTP(string secret, long timestamp)
    {
        var key = Encoding.UTF8.GetBytes(secret);
        var msg = BitConverter.GetBytes(timestamp);
        if (BitConverter.IsLittleEndian)
            Array.Reverse(msg);

        using (var hmac = new HMACSHA1(key))
        {
            var hash = hmac.ComputeHash(msg);
            int offset = hash[hash.Length - 1] & 0xf;
            int code = (hash[offset] & 0x7f) << 24 |
                       (hash[offset + 1] & 0xff) << 16 |
                       (hash[offset + 2] & 0xff) << 8 |
                       (hash[offset + 3] & 0xff);
            return (code % 1000000).ToString("D6");
        }
    }

    public void Clone(string remoteUrl)
    {
        Console.WriteLine($"Cloning from {remoteUrl}");
        // Simplified clone - would pull repo from remote in production
        Init();
    }
}

class Omi
{
    static void Main(string[] args)
    {
        if (args.Length == 0)
        {
            PrintHelp();
            return;
        }

        var settings = new Settings();
        settings.LoadFromFile("settings.txt");

        var dbPath = settings.LocalRepoPath ?? "repo.omi";
        var repo = new OmiRepository(settings, dbPath);

        var command = args[0].ToLower();

        try
        {
            switch (command)
            {
                case "init":
                    repo.Init();
                    break;

                case "add":
                    if (args.Length < 2)
                    {
                        Console.WriteLine("Usage: omi add <file>");
                        return;
                    }
                    if (args[1] == "--all")
                    {
                        var files = Directory.GetFiles(".", "*", SearchOption.AllDirectories);
                        foreach (var f in files)
                        {
                            if (!f.Contains(".omi"))
                                repo.AddFile(f);
                        }
                    }
                    else
                    {
                        repo.AddFile(args[1]);
                    }
                    break;

                case "commit":
                    if (args.Length < 3 || args[1] != "-m")
                    {
                        Console.WriteLine("Usage: omi commit -m \"message\"");
                        return;
                    }
                    repo.Commit(args[2]);
                    break;

                case "push":
                    Console.Write("OTP code: ");
                    var otpPush = ReadOTP();
                    repo.Push(otpPush);
                    break;

                case "pull":
                    Console.Write("OTP code: ");
                    var otpPull = ReadOTP();
                    repo.Pull(otpPull);
                    break;

                case "list":
                    repo.ListRepositories();
                    break;

                case "status":
                    repo.ShowStatus();
                    break;

                case "log":
                    repo.ShowLog();
                    break;

                case "clone":
                    if (args.Length < 2)
                    {
                        Console.WriteLine("Usage: omi clone <url>");
                        return;
                    }
                    repo.Clone(args[1]);
                    break;

                default:
                    Console.WriteLine($"Unknown command: {command}");
                    PrintHelp();
                    break;
            }
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Error: {ex.Message}");
        }
    }

    static void PrintHelp()
    {
        Console.WriteLine(@"
Omi - Git-like version control for retro and modern systems

Usage: omi <command> [options]

Commands:
  init              Initialize a new repository
  add <file>        Stage a file for commit
  add --all         Stage all files
  commit -m <msg>   Create a commit with message
  push              Push to remote repository (requires OTP)
  pull              Pull from remote repository (requires OTP)
  clone <url>       Clone a remote repository
  list              List all repositories
  status            Show repository status
  log               Show commit history

Configuration: Edit settings.txt to configure server URL and API key

Examples:
  omi init
  omi add file.txt
  omi commit -m ""First commit""
  omi log
");
    }

    static string ReadOTP()
    {
        var sb = new StringBuilder();
        while (true)
        {
            var key = Console.ReadKey(true);
            if (key.Key == ConsoleKey.Enter)
                break;
            if (key.Key == ConsoleKey.Backspace && sb.Length > 0)
                sb.Length--;
            else if (char.IsDigit(key.KeyChar))
                sb.Append(key.KeyChar);
        }
        Console.WriteLine();
        return sb.ToString();
    }
}
